libopenage/unit/action.h

```diff
@@ -8,13 +8,77 @@
 #include "../pathfinding/path.h"
 #include "../gamestate/resource.h"
 #include "attribute.h"
+#include "research.h"
 #include "unit.h"
 #include "unit_container.h"
 
 namespace openage {
 
 class TerrainSearch;
 
+// TODO use a type instead of unsigned int for time
+
+/**
+ * A interval triggering timer used in actions.
+ * TODO find a better name for triggers
+ */
+class IntervalTimer {
+public:
+
+	/**
+	 * Constructs a timer with a given interval
+	 */
+	IntervalTimer(unsigned int interval);
+
+	/**
+	 * Constructs a timer with a given interval which will
+	 * stop after a given number of triggers.
+	 */
+	IntervalTimer(unsigned int interval, int max_triggers);
+
+	void skip_to_trigger();
+
+	bool update(unsigned int time);
+
+	/**
+	 * Returns the time until the next trigger
+	 */
+	unsigned int get_time_left() const;
+
+	float get_progress() const;
+
+	/**
+	 * Returns true if at least one interval has passed.
+	 */
+	bool has_triggers() const;
+
+	/**
+	 * Returns true if the interval passed have reached the max.
+	 */
+	bool finished() const;
+
+	/**
+	 * Returns the number of intervals passed.
+	 */
+	int get_triggers() const { return this->triggers; }
+
+	unsigned int get_interval() const { return this->interval; }
+
+	void set_interval(unsigned int interval) { this->interval = interval; }
+
+private:
+
+	unsigned int interval;
+
+	int max_triggers;
+
+	unsigned int time_left;
+
+	int triggers;
+
+};
```

Comment:
I think this should go into a separate file. All those monster files deserve to be split up, actually :)

This timer triggering should be synched with @Tomatower's event triggers, where such a timer would be registered and you'd get a function call for the right time. Might be a very good start for integrating into the event-architecture.

Reply:
@Tomatower code is far away and a huge changes will have to be made in order to be able to move into curves (I will write more at his pr)

Reply:
@TheJJ I'm not making a new file because the timer is a stepping stone to transition better to Events or Curves (I prefer Curves, more at #744)
(no problem if you insist)

Reply:
The events are actually designed for that, and the only thing that is missing is a nice Api in the game engine - hopefully without major changes to the existing codebase - so inspire me for the best possbile api 😃